broker/src/main/java/io/camunda/zeebe/broker/partitioning/RaftPartitionFactory.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.broker.partitioning;
+
+import io.atomix.primitive.partition.PartitionMetadata;
+import io.atomix.raft.partition.RaftPartition;
+import io.atomix.raft.partition.RaftPartitionConfig;
+import io.atomix.raft.partition.RaftStorageConfig;
+import io.atomix.raft.storage.log.DelayedFlusher;
+import io.atomix.raft.storage.log.RaftLogFlusher;
+import io.camunda.zeebe.broker.Loggers;
+import io.camunda.zeebe.broker.raft.ZeebeEntryValidator;
+import io.camunda.zeebe.broker.system.configuration.BrokerCfg;
+import io.camunda.zeebe.broker.system.configuration.ExperimentalCfg;
+import io.camunda.zeebe.broker.system.configuration.RaftCfg.FlushConfig;
+import io.camunda.zeebe.snapshots.ReceivableSnapshotStoreFactory;
+import io.camunda.zeebe.util.FileUtil;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Paths;
+import java.time.Duration;
+
+public final class RaftPartitionFactory {
+  public static final String GROUP_NAME = "raft-partition";
+
+  private final BrokerCfg brokerCfg;
+  private final ReceivableSnapshotStoreFactory snapshotStoreFactory;
+
+  public RaftPartitionFactory(
+      final BrokerCfg brokerCfg, final ReceivableSnapshotStoreFactory snapshotStoreFactory) {
+    this.brokerCfg = brokerCfg;
+    this.snapshotStoreFactory = snapshotStoreFactory;
+  }
+
+  public RaftPartition createRaftPartition(final PartitionMetadata partitionMetadata) {
+    final var storageConfig = new RaftStorageConfig();
+    final var partitionConfig = new RaftPartitionConfig();
+
+    final var maxMessageSize = brokerCfg.getNetwork().getMaxMessageSizeInBytes();
+    final var segmentSize = brokerCfg.getData().getLogSegmentSizeInBytes();
+    if (segmentSize < maxMessageSize) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Expected the raft segment size greater than the max message size of %s, but was %s.",
+              maxMessageSize, segmentSize));
+    }
+    storageConfig.setSegmentSize(segmentSize);
+    final var dataDirectory = Paths.get(brokerCfg.getData().getDirectory()).resolve(GROUP_NAME);
+
+    try {
+      FileUtil.ensureDirectoryExists(dataDirectory);
+    } catch (final IOException e) {
+      throw new UncheckedIOException("Failed to create Raft data directory", e);
+    }
+
+    storageConfig.setDirectory(dataDirectory.toString());
+    storageConfig.setPersistedSnapshotStoreFactory(snapshotStoreFactory);
```

Comment:
:thought_balloon: Since now the config is built per partition, we can set the partition's directory here instead of the root directory for all partitions. Similary, if we built the snapshotStore already, we can pass it here instead of the factory.

Reply:
Actually, `storageConfig.directory` is not used anywhere. You can remove it.